opentelemetry-sdk/src/metrics/internal/mod.rs

```diff
@@ -135,6 +134,66 @@ where
             );
         }
     }
+
+    /// Iterate through all attribute sets and populate `DataPoints` in readonly mode.
+    pub(crate) fn collect_readonly<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
+    where
+        MapFn: FnMut(Vec<KeyValue>, &A) -> Res,
+    {
+        prepare_data(dest, self.count.load(Ordering::SeqCst));
+        if self.has_no_attribute_value.load(Ordering::Acquire) {
+            dest.push(map_fn(vec![], &self.no_attribute_tracker));
+        }
+
+        let Ok(trackers) = self.trackers.read() else {
+            return;
+        };
+
+        let mut seen = HashSet::new();
+        for (attrs, tracker) in trackers.iter() {
+            if seen.insert(Arc::as_ptr(tracker)) {
+                dest.push(map_fn(attrs.clone(), tracker));
+            }
+        }
+    }
+
+    /// Iterate through all attribute sets, populate `DataPoints` and and reset.
+    pub(crate) fn collect_and_reset<Res, MapFn>(&self, dest: &mut Vec<Res>, mut map_fn: MapFn)
+    where
+        MapFn: FnMut(Vec<KeyValue>, A) -> Res,
+    {
+        prepare_data(dest, self.count.load(Ordering::SeqCst));
+        if self.has_no_attribute_value.swap(false, Ordering::AcqRel) {
+            dest.push(map_fn(
+                vec![],
+                self.no_attribute_tracker.clone_and_reset(&self.config),
+            ));
+        }
+
+        let trackers = match self.trackers.write() {
```

Comment:
Something we can come back in the future - the current behavior of aggressively cleaning up valuemap during every collect() is not really needed - we only need to remove those trackers which did not have any new update since the last collect. Doing this requires additional bookkeeping costs, but that might be a better overall design.
Anyway, not for this PR!

Reply:
I probably understand your idea, but I'm afraid that this might require holding write lock for a long time. In this case, write lock is super fast, just replace variable basically.

Reply:
Resolving now. https://github.com/open-telemetry/opentelemetry-rust/pull/2267/files#r1828334335 will make it more non-trivial than before, but will come back to this.